crates/nargo/src/cli/test_cmd.rs

```diff
@@ -0,0 +1,86 @@
+use noirc_driver::{CompiledProgram, Driver};
+use noirc_frontend::node_interner::FuncId;
+use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+use crate::{errors::CliError, resolver::Resolver};
+
+use super::add_std_lib;
+
+pub(crate) fn run(args: ArgMatches) -> Result<(), CliError> {
+    let args = args.subcommand_matches("test").unwrap();
+    let test_name = args.value_of("test_name").unwrap_or("");
+    let allow_warnings = args.is_present("allow-warnings");
+    run_tests(test_name, allow_warnings)
+}
+
+fn run_tests(test_name: &str, allow_warnings: bool) -> Result<(), CliError> {
+    let backend = crate::backends::ConcreteBackend;
+
+    let package_dir = std::env::current_dir().unwrap();
+    let mut driver = Resolver::resolve_root_config(&package_dir, backend.np_language())?;
+    add_std_lib(&mut driver);
+
+    if driver.check_crate(allow_warnings).is_err() {
+        std::process::exit(1);
+    }
+
+    let test_functions = driver.get_all_test_functions_in_crate_matching(test_name);
+    println!("Running {} test functions...", test_functions.len());
+    let mut failing = 0;
+
+    let writer = StandardStream::stderr(ColorChoice::Always);
+    let mut writer = writer.lock();
+
+    for test_function in test_functions {
+        let test_name = driver.function_name(test_function);
+        write!(writer, "Testing {test_name}... ").expect("Failed to write to stdout");
+        writer.flush().ok();
+
+        match compile_test(test_name, test_function, &driver, allow_warnings) {
+            Ok(_) => {
+                writer.set_color(ColorSpec::new().set_fg(Some(Color::Green))).ok();
+                writeln!(writer, "ok").ok();
+            }
+            Err(_) => {
+                // An error likely was printed in the meantime, so this is the start of a newline.
+                // Add an extra newline as well to help separate failing tests.
+                write!(writer, "{test_name} ").ok();
+                writer.set_color(ColorSpec::new().set_fg(Some(Color::Red))).ok();
+                println!("failed\n");
+                failing += 1;
+            }
+        }
+        writer.reset().ok();
+    }
+
+    if failing == 0 {
+        writer.set_color(ColorSpec::new().set_fg(Some(Color::Green))).unwrap();
+        writeln!(writer, "All tests passed").ok();
+    } else {
+        let plural = if failing == 1 { "" } else { "s" };
+        writer.set_color(ColorSpec::new().set_fg(Some(Color::Red))).unwrap();
+        writeln!(writer, "{failing} test{plural} failed").ok();
+        std::process::exit(1);
+    }
+
+    writer.reset().ok();
+    Ok(())
+}
+
+fn compile_test(
+    test_name: &str,
+    main: FuncId,
+    driver: &Driver,
+    allow_warnings: bool,
+) -> Result<CompiledProgram, CliError> {
+    let backend = crate::backends::ConcreteBackend;
+    let language = backend.np_language();
+
+    driver
+        .compile_no_check(language, false, allow_warnings, Some(main))
```

Comment:
I think we have to do a `backend.solve` here. Although a test has no inputs and everything would be constant, during acir gen these constants are assigned witness indices and their evaluation is held off until the PWG. 

For example, if we used the logic inside of the `pedersen_check` and provided incorrect values our test should fail, but `nargo test` says our test passes:
```
#[test]
fn testing() {
    let x = 0;
    let y = 1;
    let salt = 42;

    let out_x = 0x229fb88be21cec523e9223a21324f2e305aea8bff9cdbcb3d0c6bba384666ea1;
    // This is the correct out_y
    // let out_y = 0x296b4b4605e586a91caa3202baad557628a8c56d0a1d6dff1a7ca35aed3029d5;
    // This out_y should cause the test to fail, but instead it passes
    let out_y = 0x50;

    let res = std::hash::pedersen([x, y]);
    constrain res[0] == out_x;
    constrain res[1] == out_y;

    let raw_data = [x,y];
    let mut state = 0;
    for i in 0..2 {
        state = state * 8 + raw_data[i];
    }
    state += salt;
    let hash = std::hash::pedersen([state]);
    constrain std::hash::pedersen([43])[0] == hash[0];
}
```
We can re-use the execute command in PR #725 in some way, but I think just placing something like this after `compile_no_check` should work:
```
    let mut solved_witness = BTreeMap::new();
    let x = backend.solve(&mut solved_witness, compiled_program.circuit.opcodes.clone())?;
    Ok(compiled_program)
```

Reply:
Good catch, looks like I originally had it correct before removing this step. I've added this solving step and reformatted output a bit. It looks like we can't generate errors that are as good during this step. The tests now print out a generic "failed" if we fail during this step.